5.6.34 ctrlr panel layer list update

dnaldoog opened the pull request:
Adds Eyes! to layer window!

👀

dnaldoog commented:
I believe I have fixed the LookAndFeel Assertion on quit:

https://forum.juce.com/t/lookandfeel-assert-firing-on-5-2-quit/24832/7

Needed changes to 16 files.

dnaldoog commented:
From what I read it **is** a potential problem - it indicates a memory leak with L&F - also I think the branch containing your latest layer code that I added the eyes to crashes in debug when closed, but I think the latest PR I did doesn’t crash. I’ll have to check when I get up.

damiensellier commented:
I'll do it if you want, it's the same method I applied to the drag arrows. It requires the svg to have fill="".

such as in : 

```
void DragIconComponent::paint(juce::Graphics& g)
{
    std::unique_ptr<juce::Drawable> icon = juce::Drawable::createFromImageData(dragDropIcon, strlen(dragDropIcon));

    if (icon)
    {
        auto iconColour = getLookAndFeel().findColour(juce::Label::textColourId);

        // This is the key change: we replace the black fill colour with the correct one. I spent half an hour on this one :(
        icon->replaceColour(juce::Colours::black, iconColour);

        icon->drawWithin(g, getLocalBounds().toFloat(), juce::RectanglePlacement::centred, 1.0f);
    }
}
```

dnaldoog commented:
Actually my branch is crashing too - Time to dig deeper.

dnaldoog commented:
I am discovering that there is an infinite loop happening during runtime, so it's a pretty serious bug. It only manifest when exiting, but it is there. Am working on it.

damiensellier commented:
Just to let you know, when running the app attached to your IDE will crash the app if you don't stop it via the IDE, it's not because it's because of the link between the 2 softwares.
So, to be considered as a real bug, you'd need to use CtrlrX as a standalone app, (not attached to visual studio or xcode in anyway) and if ever when closing it still have the freeze that's an issue.

For exemple if I compile ctrlrx in xcode with the automatic launch after success, and it launches CtrlrX after compilation, if I close CtrlrX from CtrlrX it will freeze most of the time, it's because of the attachement to the IDE.

dnaldoog commented:
Oh I see. So I shouldn’t worry about it too much in that case?

by the way I think the restore button was not showing (restring) when closing layer window and reopening but I fixed it by tracking a new ID

dnaldoog commented:
Eye icon now changes colour!

<img width="795" height="231" alt="{3B0B8C2A-88E6-49E9-B711-E313D11ECC21}" src="https://github.com/user-attachments/assets/aa43f8f9-8865-4c08-8dce-275308fc2f9b" />

dnaldoog commented:
Some of the functionality of the layer window was broken, but I believe I have fixed it all now.

dnaldoog commented:
Yes I can imagine it's difficult. I have never tried doing a merge, but sometimes I just use a diff program like Beyond Compare or Meld, but still not so easy. I was going to say I only changed a few files, but to fix that bug was fairly comprehensive, so there are in fact many changes.

If you just accept the whole file and merge it against `5.6.34-Maintainance` will that work, or do you need to merge it with the `5.6.34-CtrlrPanelLayerList-UPDATE` branch then merge with `5.6.34-Maintainance`?

damiensellier commented:
> Yes I can imagine it's difficult. I have never tried doing a merge, but sometimes I just use a diff program like Beyond Compare or Meld, but still not so easy. I was going to say I only changed a few files, but to fix that bug was fairly comprehensive, so there are in fact many changes.
> 
> If you just accept the whole file and merge it against `5.6.34-Maintainance` will that work, or do you need to merge it with the `5.6.34-CtrlrPanelLayerList-UPDATE` branch then merge with `5.6.34-Maintainance`?

I'll never do a merge commit like this, because I want to be able to follow what's been modded over time in a linear way. it's useful when looking at the history for each file. 

So i'll spend some time to look carefully to what you changed, reverted etc

I also noticed you removed the ModalCallback class and a few other functions. Are you sure about it? because everything was working great on my side yesterday and everything was put together carefully.

damiensellier commented:
I'm checking what you updated, and it seems that you reverted to a older version of the Lua settings panel, I fixed everything so that all highlight colours work and everything is gone in your PR with the latest commit.
Just look at the preview in the sample code window, I write something with every highlight case to show the colours applied. but in your PR the sample code is the super short one from ctrlr 5.3.x

https://github.com/damiensellier/CtrlrX/commit/335419ef884f25844e756547506f7f82f87e70f4
`
codeDocument.replaceAllContent("-- This is a comment\nfunction myFunction(argument)\n\tcall(\"string\")\nend");
`

vs

```
	// Sample code for codeDocument
	codeDocument.replaceAllContent(
    "local function test(arg1)\n"
    "    if true then\n"
    "        return 123.456\n"
    "    end -- Comment\n"
    "    local x = 'A string'\n"
	"    local op = 1 + 2 * 3 / 4 - 5\n"
    "    local t = {1, 2, 3};\n"
    "    local err = 123_ -- Error\n"
    "end -- Enjoy!"
	);
```

and so on.


It seems that you want commented a lot of components in the destructors, I saw some that really needs to be deleteAndZero() 

As always it's hard for me to track down a bug that I don't experience on my system. If you can give me more details about the reason you went to the hunt to the forgotten component to destruct, that would help me, because from a blank panel that I build up step by step, save, reopen etc I don't get the crash or hang when closing.
The only one I had was a problem with the notifier (green or red) at the bottom when closing and its timer not over. I shortened the display time to 2500ms and added it to the destructor et voilà. 

Let me know ;)
Have a nice week

Damien

damiensellier commented:
Don't worry John! 
I'll do exactly what you suggested. I'll update step by step with the eye icon first and you'll be able to test at every step.

dnaldoog commented:
Okay, but isn't it nearly sunrise over there?

damiensellier commented:
Excellent! I'll update next week. Thanks for all those fixes.

dnaldoog commented:
>I'm checking what you updated, and it seems that you reverted to a older version of the Lua settings panel, I fixed everything so that all highlight colours work and everything is gone in your PR with the latest commit.
Just look at the preview in the sample code window, I wrote something to cover every highlight case and show the colours applied. but in your PR the sample code is the super short one from ctrlr 5.3.x

Hi Damien,

The following code changes should implement the layer visibility code and edit/restore layer functionality:
This is all the code I had added to the PR with the wrong version.

### (1) Add this code to the end of CtrlrPanel.cpp

````
void CtrlrPanel::saveLayerVisibilityStates() // Added v5.6.34. Thanks to @dnaldoog. Required for layer management
{
	// Clear any previous states
	layerVisibilityBackup.clear();
	// Get the canvas from the editor
	CtrlrPanelCanvas* canvas = getCanvas();

	if (canvas)
	{
		// Loop through all layers and save their current visibility state
		for (int i = 0; i < canvas->getNumLayers(); ++i)
		{
			CtrlrPanelCanvasLayer* layer = canvas->getLayerFromArray(i);
			if (layer)
			{
				layerVisibilityBackup.add(layer->getProperty(Ids::uiPanelCanvasLayerVisibility, true));
			}
		}
	}
}

void CtrlrPanel::restoreLayerVisibilityStates() // Added v5.6.34. Thanks to @dnaldoog. Required for layer management
{
	// Get the canvas from the editor
	CtrlrPanelCanvas* canvas = getCanvas();

	if (canvas && hasLayerVisibilityStates())
	{
		// Loop through layers and restore visibility from the backup
		for (int i = 0; i < canvas->getNumLayers() && i < layerVisibilityBackup.size(); ++i)
		{
			CtrlrPanelCanvasLayer* layer = canvas->getLayerFromArray(i);
			if (layer)
			{
				layer->setProperty(Ids::uiPanelCanvasLayerVisibility, layerVisibilityBackup.getUnchecked(i));
			}
		}
	}
	// Clear the backup after restoring
	layerVisibilityBackup.clear();
}
````
### (2) Add this code to CtrlrPanel.h
````
#include "CtrlrPanel/CtrlrPanelCanvasLayer.h" // Added v5.6.34

public:

		void saveLayerVisibilityStates(); // Added v5.6.34
		void restoreLayerVisibilityStates(); // Added v5.6.34
		juce::Array<juce::var> layerVisibilityBackup; // Added v5.6.34
		bool hasLayerVisibilityStates() const { return layerVisibilityBackup.size() > 0; }; // Added v5.6.34

````

### (3) Add this to CtrlrIDs.h

 DECLARE_ID (uiPanelCanvasLayerIsIsolated); // @line 138


### (4) Add this to constructor of CtrlrEditor.cpp

````diff
CtrlrEditor::CtrlrEditor(CtrlrProcessor* _ownerFilter, CtrlrManager& _owner)
    : AudioProcessorEditor(_ownerFilter),
    ownerFilter(_ownerFilter),
    owner(_owner),
    resizer(this, 0),
    tempResult(Result::ok()),
    menuHandlerCalled(false),
+   currentLookAndFeel(nullptr),
    lastCommandInvocationMillis(0),
+   previousGlobalLookAndFeel(nullptr)
{
    // Initialize currentLookAndFeel to a default LookAndFeel_V4 Light.
    // This provides a starting point before properties are loaded.
+   previousGlobalLookAndFeel = &LookAndFeel::getDefaultLookAndFeel();
    currentLookAndFeel = new LookAndFeel_V4(juce::LookAndFeel_V4::getLightColourScheme());
    setLookAndFeel(currentLookAndFeel); // Se
...
````
 Add to CtrlrEditor.h
 
````diff
private:
        // Use a ScopedPointer to manage the current LookAndFeel object
+ ScopedPointer<LookAndFeel> currentLookAndFeel;
+ LookAndFeel* previousGlobalLookAndFeel;

````
### (5) Destructor code for CtrlrEditor.cpp
````diff
CtrlrEditor::~CtrlrEditor()
{
- menuBar->setLookAndFeel(nullptr); // Detach LookAndFeel from menuBar first
+    // 1. First detach LookAndFeel from all components
+    setLookAndFeel(nullptr);
+    if (menuBar)
+    {
+        menuBar->setLookAndFeel(nullptr);
+    }

+    // 2. Restore the original global LookAndFeel
+    if (previousGlobalLookAndFeel != nullptr)
+    {
+        LookAndFeel::setDefaultLookAndFeel(previousGlobalLookAndFeel);
+    }

+    // 3. Now it's safe to delete the menuBar and our custom LookAndFeel
    deleteAndZero(menuBar);
+    currentLookAndFeel = nullptr; // ScopedPointer will delete automatically

    masterReference.clear();
}

````

### (6)
````diff
void CtrlrEditor::setEditorLookAndFeel(const String& lookAndFeelDesc...
-  LookAndFeel::setDefaultLookAndFeel(currentLookAndFeel); 
    }
````
### (7)
````diff
CtrlrGroup::CtrlrGroup(CtrlrModulator& owner)
-	: CtrlrComponent(owner), content(*this),
-	label (0)
+ : CtrlrComponent(owner), content(*this),
{

```


### (8)
````diff
CtrlrGroup::~CtrlrGroup()
{
	//[Destructor_pre]. You can add your own custom destruction code here..
	owner.getModulatorTree().removeListener(this);
	componentTree.removeListener(this);
	//[/Destructor_pre]

-	deleteAndZero (label);

	//[Destructor]. You can add your own custom destruction code here..
	//[/Destructor]
}

````


###  (9) CtrlrGroup.h
````diff
+ juce::ScopedPointer<Label> label;
- Label* label;
````


### (10)
````diff
CtrlrLCDLabel::CtrlrLCDLabel (CtrlrModulator &owner)
-    : CtrlrComponent(owner),
-      ctrlrLabel (0)
+	  : CtrlrComponent(owner)
````
{

### (11)
````diff
CtrlrLCDLabel::~CtrlrLCDLabel()
{
...
- deleteAndZero (ctrlrLabel);
````
### (12) CtrlLCDLabel.h

````diff
+ ScopedPointer<Label> ctrlrLabel;
- Label* ctrlrLabel;
````

### (13)
````diff
CtrlrLabel::~CtrlrLabel()
{
...
- deleteAndZero (ctrlrLabel);
````
### (14) CtrlLabel.h

````diff
+ ScopedPointer<Label> ctrlrLabel;
- Label* ctrlrLabel;
````
### (15)
````diff

CtrlrPanelCanvasLayer::CtrlrPanelCanvasLayer(CtrlrPanelCanvas &_owner)
	: layerTree(Ids::uiPanelCanvasLayer), owner(_owner), mousePass(false)
{
	layerTree.addListener (this);
	int64 data = Time::getCurrentTime().getHighResolutionTicks();
	setProperty (Ids::uiPanelCanvasLayerName, "New layer");
	setProperty (Ids::uiPanelCanvasLayerUid, Uuid ((const uint8 *)&data).toString());
	setProperty (Ids::uiPanelCanvasLayerColour, "0x000000");
	setProperty (Ids::uiPanelCanvasLayerVisibility, true);
	setProperty (Ids::uiPanelCanvasLayerIndex, 0);
+   setProperty(Ids::uiPanelCanvasLayerIsIsolated, false);
	owner.getCanvasTree().addChild (layerTree, -1, 0);

````
### (16) CtrlrPanelEditor.cpp
````diff

+CtrlrPanelNotifier::~CtrlrPanelNotifier()
+{
+    text = nullptr; // Force ScopedPointer cleanup
+}

````
in header file
````diff
-        ~CtrlrPanelNotifier() {}
+        ~CtrlrPanelNotifier();
````
### (17)
````diff

CtrlrPanelEditor::CtrlrPanelEditor(CtrlrPanel& _owner, CtrlrManager& _ctrlrManager, const String& panelName)
    : Component(L"Ctrlr Panel Editor"),
    lastEditMode(true),
    ctrlrManager(_ctrlrManager),
    owner(_owner),
    panelEditorTree(Ids::uiPanelEditor),
    ctrlrComponentSelection(nullptr),
    ctrlrPanelProperties(nullptr),
   spacerComponent(nullptr),
    lookAndFeel(nullptr),
+    previousGlobalLookAndFeel(nullptr)
{
+    previousGlobalLookAndFeel = &LookAndFeel::getDefaultLookAndFeel();

````
### (18)
````diff

CtrlrPanelEditor::~CtrlrPanelEditor()
{
- componentAnimator.removeChangeListener (this); 
+    // 1. Detach from specific components
+    setLookAndFeel(nullptr);
+    if (getCanvas())
+    {
+        getCanvas()->setLookAndFeel(nullptr);
+    }

+    // 2. Restore the previous global LookAndFeel (if we changed it)
+    if (previousGlobalLookAndFeel != nullptr)
+    {
        LookAndFeel::setDefaultLookAndFeel(previousGlobalLookAndFeel);
+    }

+    // 3. Now the lookAndFeel unique_ptr can safely delete the object

+    // 4. Continue with cleanup
    getPanelEditorTree().removeListener(this);
    owner.getPanelTree().removeListener(this);
    owner.getPanelTree().removeChild(getPanelEditorTree(), 0);
    ctrlrComponentSelection->removeChangeListener(ctrlrPanelProperties);
    masterReference.clear();
    deleteAndZero(ctrlrPanelProperties);
    deleteAndZero(spacerComponent);
    deleteAndZero(ctrlrPanelViewport);
}

````
### (19) CtrlrPanelEditor.h
````diff
+ #include "CtrlrPanelEditorIcons.h"



+class CtrlrPanelNotifier : public Component
+{
+public:

- class CtrlrPanelNotifier : public Component,
-                           public LookAndFeel_V4
-                             // Added back v5.6.31 for file management bottom notification bar
-{
	
	class CtrlrPanelEditor  :	public Component,
							public ValueTree::Listener,
							public CtrlrLuaObject,
                            public LookAndFeel_V4,
-                            public juce::ChangeListener // Add this line


...
// at end of file
	std::unique_ptr<juce::LookAndFeel> lookAndFeel;
+	juce::LookAndFeel* previousGlobalLookAndFeel;
};
````
### (20)
Add CtrlrPanelEditorIcons.h
Add CtrlrPanelEditorIcons.cpp

to project in :
`Source/UIComponents/CtrlrPanel`
### (21)
Copy attached CtrlrLayerList.cpp/.h files into project (see attached zip file)
### (22)
Copy attached CtrlrLayerListItem.cpp/.h files into project (see attached file)


[layer visibility files.zip](https://github.com/user-attachments/files/22084803/layer.visibility.files.zip)

damiensellier commented:
Hi, I'm updating the layer editor in a way we get the icons updating their colour related to the current colourScheme right away (no need to close and reopen the layer window). It was not the case with this PR code. 

I'll do the crash fix in another commit .

<img width="816" height="567" alt="Capture d’écran, le 2025-09-02 à 19 02 28" src="https://github.com/user-attachments/assets/4e8d5d8d-2108-4b06-b2c7-53a4efdb6048" />

dnaldoog commented:
The crash code and layer code is contained in this branch ( using everything above) ☝️ 
- https://github.com/dnaldoog/CtrlrX/tree/layerplusnoncrashcode

damiensellier commented:
Perfect, I'll look into it tonight and also do the update for the resource panel to show the image slider image previews.

dnaldoog commented:
In Windows you don’t have to close and reopen the layer dialogue in order to see theme changes. It seems to work fine as it is must be something to do with macOS.

damiensellier commented:
UPDATED on 2025.09.02 for CtrlrX 5.6.34 commit 6f26ef3642e76f1abe53e242bd5de52325bef1da

I'll merge this branch to the main one https://github.com/damiensellier/CtrlrX/tree/5.6.34-Maintenance

damiensellier commented:
Don't worry, it's all good. Next!

dnaldoog commented:
The code for changing the eye color is contained in CtrlrPanelEditor.cpp

https://github.com/dnaldoog/CtrlrX/blob/5289784d2fee4f7b1a04d60a7e3ef828e5829ce1/Source/UIComponents/CtrlrPanel/CtrlrPanelEditorIcons.cpp#L125
